### PR TITLE
Migrate artifact hosting to cloudsmith

### PR DESCRIPTION
**Pull request by krishnangovindraj**
## What is the goal of this PR?
Updates artifact deployment & consumption rules to use cloudsmith instead of the self-hosted sonatype repository.

## What are the changes implemented in this PR?
* Updates deployment rules & scripts to use Cloudsmith.
* Updates consumption rules to use Cloudsmith instead of nexus.
* Updates documentation relating to repo.vaticle.com

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -651,36 +627,6 @@ Assemble files for HashiCorp Packer deployment
...
-| <a id="assemble_rpm-workspace_refs"></a>workspace_refs |  JSON file with other Bazel workspace references   |  <code>None</code> |
-| <a id="assemble_rpm-installation_dir"></a>installation_dir |  directory into which .rpm package is unpacked at installation   |  <code>None</code> |
-| <a id="assemble_rpm-archives"></a>archives |  Bazel labels of archives that go into .rpm package   |  <code>[]</code> |
-| <a id="assemble_rpm-empty_dirs"></a>empty_dirs |  list of empty directories created at package installation   |  <code>[]</code> |
-| <a id="assemble_rpm-files"></a>files |  mapping between Bazel labels of archives that go into .rpm package and their resulting location on .rpm package installation   |  <code>{}</code> |
-| <a id="assemble_rpm-permissions"></a>permissions |  mapping between paths and UNIX permissions   |  <code>{}</code> |
-| <a id="assemble_rpm-symlinks"></a>symlinks |  mapping between source and target of symbolic links created at installation   |  <code>{}</code> |
-| <a id="assemble_rpm-tags"></a>tags |  additional tags passed to all wrapped rules   |  <code>[]</code> |
-
-
```
Comment (krishnangovindraj) [inline README.md:683]: Removes rpm targets

File: artifact/rules.bzl [hunk 2]
```diff
@@ -165,7 +168,7 @@ def artifact_file(name,
 
     http_file(
         name = name,
-        urls = ["{}/{}/{}/{}".format(repository_url, group_name, version, artifact_name)],
+        urls = ["{}/names/{}/versions/{}/{}".format(repository_url.rstrip("/"), group_name, version, artifact_name)],
```
Comment (krishnangovindraj) [inline artifact/rules.bzl:171]: Broke if the url had a trailing `/`

File: artifact/rules.bzl [hunk 3]
```diff
@@ -51,11 +51,11 @@ def _deploy_artifact_impl(ctx):
     symlinks = {
         'VERSION': version_file,
     }
-
+    cloudsmith_lib_files = ctx.attr._cloudsmith_pylib[DefaultInfo].default_runfiles.files.to_list()
     return DefaultInfo(
         executable = _deploy_script,
         runfiles = ctx.runfiles(
-            files = files,
+            files = files + cloudsmith_lib_files,
```
Comment (krishnangovindraj) [inline artifact/rules.bzl:58]: How we add the dependency on the `//common/cloudsmith:clodsmith-wrapper` wrapper to the script.

File: common/cloudsmith/BUILD
Comment (krishnangovindraj) [inline common/cloudsmith/BUILD:1]: I have no idea why it detected this as a move. Logically, it's just a new file

File: crates/CrateDeployer.kt [hunk 4]
```diff
@@ -99,7 +99,7 @@ class CrateDeployer : Callable<Unit> {
     private fun httpPut(url: String, token: String, content: ByteArray): HttpResponse {
         return NetHttpTransport()
             .createRequestFactory()
-            .buildPutRequest(GenericUrl(url), ByteArrayContent("application/json", content))
+            .buildPutRequest(GenericUrl(url), ByteArrayContent(null, content)) // TODO: Verify it works with crates.io
```
Comment (krishnangovindraj) [inline crates/CrateDeployer.kt:102]: This is a TODO I don't know how to verify.

File: helm/rules.bzl
Comment (krishnangovindraj) [inline helm/rules.bzl:1]: Roll our own `deploy_helm` rule. We used to use one from a 3rd party (we continue to use their other helm rules)

**Review by krishnangovindraj — COMMENTED** [review]
The big IF here is whether the `crates.io` expects the "application/json" header. Cloudsmith rejects it, so we're no longer setting it.

File: apt/rules.bzl [hunk 5]
```diff
@@ -235,3 +237,22 @@ deploy_apt = rule(
     Select deployment to `snapshot` or `release` repository with `bazel run //:some-deploy-apt -- [snapshot|release]
     """
 )
+
+def deploy_apt(name, target, snapshot, release, **kwargs):
```
Comment (flyingsilverfin) [inline apt/rules.bzl:241]: This should be able to either deploy to cloudsmith apt (if we use the `cloudsmith://...` convention you established) or to the previous version using curl.

Same applies to Artifact and anything else using `cloudsmith://...` endpoint.

The rules should still able to interchange with snapshot (say Docker) and release (DockerHub) right?

File: helm/rules.bzl [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+
+load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_file")
+
+def _deploy_helm_impl(ctx):
```
Comment (flyingsilverfin) [inline helm/rules.bzl:22]: This helm rule being specific to Cloudsmith is OK I figure

**Review by flyingsilverfin — APPROVED** [review]

**Pull request merged**